### DISCUSSION TITLE
CSS Strategy

### DISCUSSION THREAD

**Discussion by thescientist13** [ProjectEvergreen/www.greenwoodjs.dev#7] [Ideas]
## Overview

Would like to explore possible CSS strategies for the new site, with a few options off the top of my head, assuming light DOM rendering
- CSS Modules
- Tailwind
- Vanilla

Some specific goals / requirements I care about are:
1. It must emit actual CSS, be it compiled or otherwise, but no CSS-in-JS
1. No nested `<style>` blocks, as this can interrupt HTML parsing and lead to greater FOUC
1. Would like to keep CSS co-located, so _header.js_ + _header.css_

### CSS Modules
There currently isn't a mechanism for [**CSS Modules**](https://github.com/css-modules/css-modules) with Greenwood, but I would certainly be open to developing it, as I like the authoring experience of vanilla CSS, but with binding / co-location to its usage.
```css
/* style.css */
.className {
  color: green;
}
```

```js
import styles from "./style.css";
// import { className } from "./style.css";

element.innerHTML = '<div class="' + styles.className + '">';
```

### Tailwind
I've used [**Tailwind**](https://tailwindcss.com/) with a couple projects and works really well with Light DOM, statically pre-rendered custom elements.  Greenwood already has support for PostCSS so this is definitely a good option as well.

### Vanilla CSS
This option could be ok, but the part I'm concenred about is co-location, as the options would have to be:
1. Global CSS file
1. Inline CSS per component

The problem with the second option, is if we have something like this
```js
export default class Header extends HTMLElement {
  connectedCallback() {
    this.innerHTML = `
      <style>
        h1 {
          color: red;
        }
      </style>

       <header>
         <h1>Welcome to Greenwood!</h1>
       </header>
    `;
  }
}
```

Which would leave a bunch of `<style>` tags throughout the HTML, which wouldn't be as performant, and would cause constant reflows / repaints on load as the HTML parser keeps hitting these tags, and could lead to a greater impact of FOUC.

**Comment by aholtzman** [comment]
With the spirt of the project to keep things as close to the standards as possible, I would vote for `CSS Modules`. Tailwind would also create more dependencies, something that has generally been avoided when possible on this journey.

**Reply by thescientist13** [reply]
Well, at the end of the day the complexity for the convenience of either solution has to come from somewhere.  While PostCSS + Tailwind is effectively an optimizing compiler just working on narrowing a massive CSS file (all the Tailwind classes) based on matching on class names in input files, CSS Modules requires a bit more work / orchestration under the hood.

As CSS Modules is just a specification (with an implementation for webpack and PostCSS) the general gist as I understand it, is that an implementation would need to handle the following:
1. Process CSS files based on knowing their usages within JS files, e.g. processing `import styles from './header.module.css';`
1. Create an object of class names in the file to an object for reference within the JS template
1. Inject that CSS into the `<head>` of the page
1. From there, rewrite / hash classnames or at least scope them
1. Stitch the classnames statically back into the JavaScript module (and in theory removing the original `import styles ...` statement to remove any runtime overhead)


So basically given something like this
```js
import styles from './header.module.css';

export default class Header extends HTMLElement {
  connectedCallback() {
    this.innerHTML = `
      <header>
        <h1 class="${styles.heading}">Welcome to Greenwood!</h1>
      </header>
    `;
  }
}

customElements.define('app-header', Header);
```

```css
.heading {
  color: blue;
  text-align: center;
}
```

The outcome / output would look something like this when it is all said and done:
1. Extract the CSS and at least do basic scoping
    ```css
   .header-heading {
      color: blue;
      text-align: center;
    }
    ```
1. Embed the CSS into the `<head>` of the document inline (or save to a file and `<link>`)
    ```html
     <head>
       <style>
         .header-heading{color:blue;text-align:center;}
       </style>
     </head>
    ```
1.  Transform the JS to inline the class names and remove the CSS Module `import` line in the JS too so as to avoid any runtime overhead
    ```js
    export default class Header extends HTMLElement {
      connectedCallback() {
        this.innerHTML = `
          <header>
            <h1 class="header-heading">Welcome to Greenwood!</h1>
          </header>
        `;
      }
    }
    
    customElements.define('app-header', Header);
    ```
1. For proper alignment / expectations with CSS Modules, would need to comply with all items captured in the [ICSS specification](https://github.com/css-modules/icss)

----

That said, I've created a draft PR ( https://github.com/ProjectEvergreen/www.greenwoodjs.dev/pull/11 ) to start playing around with it and made some decent progress after only hacking on it for a few hours, so definitely take a look and see what you think.  So far this has involved leveraging three ASTs / APIs for parsing content including html-parser, css-tree, and acorn (+ acorn-walk), which are technically already deps of `@greenwood/cli` so... not sure how to score that?  😅  🤷‍♂️ 

Either way, this will definitely be a plugin so if users feel they are OK with opting into that overhead for that workflow they can, which is more or less the same investment as using our PostCSS plugin + Tailwind, so low friction whichever path you take.  🤞 
 
Will plan to review / demo during our next weekly meeting.  👍

**Comment by thescientist13** [comment]
Perhaps this could more align with the vanilla or CSS Modules approaches, but it might be cool to leverage [**Open Props**](https://open-props.style/) as our (sub) atomic utility library?

**Comment by thescientist13** [comment]
Final strategy documented here - https://github.com/ProjectEvergreen/greenwood/pull/1234

**Comment by thescientist13** [comment]
Ok, we're locked in, here goes nothing!  🤞
